### PR TITLE
A function to report scaling issues

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
This PR adds a function to report potential scaling issues in a model:

```
import pyomo.environ as pe
from pyomo.util.report_scaling import report_scaling
import logging

logging.basicConfig(level=logging.INFO)

m = pe.ConcreteModel()
m.x = pe.Var(list(range(5)))
m.c = pe.Constraint(list(range(3)))

m.x[0].setlb(-1)
m.x[0].setub(1)
m.x[1].setlb(1e5)
m.x[1].setub(1e7)
m.x[2].setlb(-1e5)
m.x[2].setub(0)
m.x[3].setlb(-20)
m.x[3].setub(20)

m.obj1 = pe.Objective(expr=1e-8*m.x[0] + pe.exp(m.x[3]) + m.x[1]*m.x[2])
m.obj2 = pe.Objective(expr=m.x[0]*m.x[3] + m.x[1]**2)

m.c[0] = m.x[0] + m.x[3] == 0
m.c[1] = 1/m.x[1] == 1
m.c[2] = m.x[1]*m.x[3] == 1

report_scaling(m)
```

Outputs

```

The following variables are not bounded. Please add bounds.
LB          UB          Var
-inf        inf         x[4]

The following variables have large bounds. Please scale them.
LB          UB          Var
1.00e+05    1.00e+07    x[1]
-1.00e+05   0.00e+00    x[2]

The following objectives have potentially large coefficients. Please scale them.
obj1
    Coef LB     Coef UB     Var
    2.06e-09    4.85e+08    x[3]
    -1.00e+05   0.00e+00    x[1]
    1.00e+05    1.00e+07    x[2]

obj2
    Coef LB     Coef UB     Var
    2.00e+05    2.00e+07    x[1]

The following objectives have small coefficients.
obj1
    Coef LB     Coef UB     Var
    1.00e-08    1.00e-08    x[0]

The following constraints have potentially large coefficients. Please scale them.
c[2]
    Coef LB     Coef UB     Var
    1.00e+05    1.00e+07    x[3]

The following constraints have small coefficients.
c[1]
    Coef LB     Coef UB     Var
    -1.00e-10   -1.00e-14   x[1]

The following constraints have bodies with large bounds. Please scale them.
LB          UB          Constraint
-2.00e+08   2.00e+08    c[2]

```

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/util/report_scaling.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
+import pyomo.environ as pe
```
Comment (mrmundt) [inline pyomo/util/report_scaling.py:1]: Needs copyright assertion

File: pyomo/util/tests/test_report_scaling.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import pyomo.environ as pe
```
Comment (mrmundt) [inline pyomo/util/tests/test_report_scaling.py:1]: Need copyright assertion

**Review by mrmundt — CHANGES_REQUESTED** [review]

**Comment by mrmundt** [comment]
I really love this utility, @michaelbynum . Nicely done!

**Comment by mrmundt** [comment]
@jsiirola - We may want to add a call to this in the trust region code. We assume correct scaling, but with this, we'd actually be able to assert it.

File: pyomo/util/report_scaling.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:11]: Done

File: pyomo/util/tests/test_report_scaling.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/util/tests/test_report_scaling.py:11]: Done.

**Comment by michaelbynum** [comment]
Thanks, @mrmundt !

File: pyomo/util/report_scaling.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    for v in var_set:
+        if v.lb is None:
+            v_lb = -math.inf
+        else:
+            v_lb = v.lb
+        if v.ub is None:
+            v_ub = math.inf
+        else:
+            v_ub = v.ub
```
Comment (jsiirola) [inline pyomo/util/report_scaling.py:53]: We repeat this a LOT.  Would a helper function be warranted:
```python
def _bounds_to_float(lb, ub):
    if lb is None:
        lb = -math.inf
    if ub is None:
        ub - math.inf
    return lb, ub
```
then, e.g.
```python
for c in var_set:
    v_lb, v_ub = _bounds_to_float(*v.bounds)
```

File: pyomo/util/tests/test_report_scaling.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        match = re.search('\n\nThe following variables are not bounded.*\s*Var\s*LB\s*UB\s*x\[4\]\s*-inf\s*inf\s*The '
+                          'following variables have large bounds.*\s*Var\s*LB\s*UB\s*x\[1\]\s*.*\s*x\[2\]\s*.*\s*.*\s*'
+                          'The following objectives have potentially large coefficients.*\s*obj1\s*Var\s*Coef LB\s*Coef '
+                          'UB\s*x\[3\]\s*.*\s*.*\s*x\[1\]\s*.*\s*.*\s*x\[2\]\s*.*\s*.*\s*obj2\s*Var\s*Coef LB\s*Coef '
+                          'UB\s*x\[1\]\s*.*\s*.*\s*The following objectives have small coefficients.*\s*obj1\s*Var\s*'
+                          'Coef LB\s*Coef UB\s*x\[0\]\s*.*\s*.*\s*The following constraints have potentially large '
+                          'coefficients.*\s*c\[2\]\s*Var\s*Coef LB\s*Coef UB\s*x\[3\]\s*.*\s*.*\s*The following '
+                          'constraints have small coefficients.*\s*c\[1\]\s*Var\s*Coef LB\s*Coef UB\s*x\[1\]\s*.*'
+                          '\s*.*\s*The following constraints have bodies with large bounds.*\s*Constraint\s*LB\s*UB\s*'
+                          'c\[2\]\s*.*\s*.*\s*', s)
```
Comment (jsiirola) [inline pyomo/util/tests/test_report_scaling.py:56]: Please reformat this as a multiline (triple-quoted) string.  Do we need it to be a regex match?  I think I would rather test against a concrete baseline...

File: pyomo/util/report_scaling.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import pyomo.environ as pe
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:11]: I think it is important to stick to that convention in documentation. I don't think it matters here. I wouldn't mind changing it though.

File: pyomo/util/report_scaling.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+def _get_longest_name(comps):
+    longest_name = 0
+
+    for i in comps:
+        i_len = len(str(i))
+        if i_len > longest_name:
+            longest_name = i_len
+
+    if longest_name > 195:
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:32]: I think I just picked 200 and subtracted 5 to account for the 5 columns of spaces added on line 43. This is just to limit the number of columns allocated for printing variable/constraint names. I'm certainly open to better ways to format the output.

File: pyomo/util/report_scaling.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        longest_name = 195
+    if longest_name < 12:
+        longest_name = 12
+
+    return longest_name
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:37]: Oh, and here is the better way :)

File: pyomo/util/report_scaling.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def _print_var_set(var_set):
+    longest_varname = _get_longest_name(var_set)
+
+    s = f'{"Var":<{longest_varname + 5}}{"LB":>12}{"UB":>12}\n'
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:43]: Genius!

File: pyomo/util/report_scaling.py (hunk 3 above)
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:53]: I like it.

File: pyomo/util/report_scaling.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        for v, der_lb, der_ub in der_bounds:
+            s += f'    {str(v):<{longest_vname + 5}}{der_lb:>12.2e}{der_ub:>12.2e}\n'
+        s += '\n'
+    return s
+
+
+def _check_coefficents(comp, expr, too_large, too_small, largs_coef_map, small_coef_map):
+    ders = reverse_sd(expr)
+    for _v, _der in ders.items():
+        if isinstance(_v, _GeneralVarData):
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:91]: Yes, I should. It will show up in the bounds on the constraint bodies.

**Comment by michaelbynum** [comment]
@jsiirola Done. Thanks for the great suggestions!

File: pyomo/util/report_scaling.py [hunk 10]
```diff
@@ -184,10 +162,9 @@ def report_scaling(m: _BlockData, too_large: float = 5e4, too_small: float = 1e-
 
     if len(cons_with_large_bounds) > 0:
         s += 'The following constraints have bodies with large bounds. Please scale them.\n'
-        longest_cname = _get_longest_name(cons_with_large_bounds)
-        s += f'{"Constraint":<{longest_cname + 5}}{"LB":>12}{"UB":>12}\n'
+        s += f'{"LB":<12}{"UB":<12}Constraint\n'
         for c, (c_lb, c_ub) in cons_with_large_bounds.items():
-            s += f'{str(c):<{longest_cname + 5}}{c_lb:>12.2e}{c_ub:>12.2e}\n'
+            s += f'{c_lb:<12.2e}{c_ub:<12.2e}{str(c)}\n'
```
Comment (jsiirola) [inline pyomo/util/report_scaling.py:167]: Would you consider making this right and not left justified?  Looking at the reference output in the tests, it wsa a bit jarring that the "e"'s didn't line up.

File: pyomo/util/report_scaling.py [hunk 11]
```diff
@@ -89,11 +71,10 @@ def _check_coefficents(comp, expr, too_large, too_small, largs_coef_map, small_c
     ders = reverse_sd(expr)
     for _v, _der in ders.items():
         if isinstance(_v, _GeneralVarData):
```
Comment (jsiirola) [inline pyomo/util/report_scaling.py:73]: When is v not a `_GeneralVarData`?

Comment (michaelbynum) [reply pyomo/util/report_scaling.py:73]: If v is not a `_GeneralVarData`, then it is a `ParamData` or some other node in the expression tree that we do not care about in this context.

File: pyomo/util/report_scaling.py (hunk 10 above)
Comment (michaelbynum) [reply pyomo/util/report_scaling.py:167]: I tried, but then I have to make the names right justified, which I think means I would need to calculate the length of the variable names?

Comment (jsiirola) [reply pyomo/util/report_scaling.py:167]: I was thinking making LB and UB right-justified and name left-justified.  We do this with the tabular writer, too (in general, right-justified looks good, but not at the end of the line, especially when that field is really long).

Comment (michaelbynum) [reply pyomo/util/report_scaling.py]: Done!

**Review by mrmundt — APPROVED** [review]

**Pull request merged**